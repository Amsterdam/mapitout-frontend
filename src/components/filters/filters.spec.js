import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import { mergeDeepLeft } from "ramda";

import ComponentFilters from "./filters";

import fixtureFilters from "../../__fixtures__/filters";

const localVue = createLocalVue();
localVue.use(Vuex);

const updateFiltersSpy = jest.fn();
const setRootFilterSpy = jest.fn();

const defaultStore = {
  modules: {
    filters: {
      namespaced: true,
      mutations: {
        setRootFilter: setRootFilterSpy
      },
      actions: {
        update: updateFiltersSpy
      },
      state: {
        rootFilter: null,
        filters: fixtureFilters
      }
    }
  }
};

const createWrapper = (options = {}) =>
  shallowMount(
    ComponentFilters,
    mergeDeepLeft(options, {
      localVue,
      stubs: ["icon-filters", "component-filter", "component-filter-group"]
    })
  );

describe("ComponentFilters", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should create", () => {
    const wrapper = createWrapper({
      store: new Vuex.Store(defaultStore)
    });

    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("should correctly process the raw filters", () => {
    const wrapper1 = createWrapper({
      store: new Vuex.Store(defaultStore)
    });

    expect(wrapper1.vm.filtersForRoot).toMatchSnapshot();

    const wrapper2 = createWrapper({
      store: new Vuex.Store(
        mergeDeepLeft(
          {
            modules: {
              filters: {
                state: {
                  rootFilter: fixtureFilters[1]
                }
              }
            }
          },
          defaultStore
        )
      )
    });

    expect(wrapper2.vm.filtersForRoot).toMatchSnapshot();
  });

  it("should not render the back button when rootFilter is set to null", () => {
    const wrapper = createWrapper({
      store: new Vuex.Store(defaultStore)
    });

    expect(wrapper.contains(".back")).toBe(false);
  });

  it("should render the root filters when clicking back", () => {
    const store = mergeDeepLeft(
      {
        modules: {
          filters: {
            state: {
              rootFilter: fixtureFilters[0]
            }
          }
        }
      },
      defaultStore
    );
    const wrapper = createWrapper({
      store: new Vuex.Store(store)
    });
    wrapper.find(".back").trigger("click");

    expect(setRootFilterSpy).toHaveBeenCalledTimes(1);
    expect(setRootFilterSpy.mock.calls[0][1]).toBe(null);
  });
});
