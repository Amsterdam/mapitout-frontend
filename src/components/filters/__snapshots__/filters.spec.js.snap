// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentFilters should correctly process the raw filters 1`] = `
Array [
  Object {
    "iconComponent": "icon-education",
    "id": 2,
    "isFilter": true,
    "name": "Schools",
    "parentId": 0,
    "propertyId": null,
    "selected": false,
    "value": "School",
  },
  Object {
    "iconComponent": "icon-bus",
    "id": 1,
    "isFilter": true,
    "name": "Stations",
    "parentId": 0,
    "propertyId": null,
    "selected": false,
    "value": "Station",
  },
]
`;

exports[`ComponentFilters should correctly process the raw filters 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "group": "School programs",
        "iconComponent": "icon-education",
        "id": 3,
        "name": "Age Range 4-11",
        "parentId": 2,
        "propertyId": 1,
        "selected": false,
        "value": "Age range 4-11",
      },
      Object {
        "group": "School programs",
        "iconComponent": "icon-education",
        "id": 4,
        "name": "Age Range 4-18",
        "parentId": 2,
        "propertyId": 2,
        "selected": false,
        "value": "Age range 4-18",
      },
      Object {
        "group": "School programs",
        "iconComponent": "icon-education",
        "id": 5,
        "name": "Primary Education",
        "parentId": 2,
        "propertyId": 3,
        "selected": false,
        "value": "Primary education",
      },
      Object {
        "group": "School programs",
        "iconComponent": "icon-education",
        "id": 6,
        "name": "Secondary Education",
        "parentId": 2,
        "propertyId": 4,
        "selected": false,
        "value": "Secondary education",
      },
    ],
    "iconComponent": "icon-education",
    "isGroup": true,
    "name": "School programs",
  },
  Object {
    "children": Array [
      Object {
        "group": "School system",
        "iconComponent": "icon-education",
        "id": 7,
        "name": "International Schools",
        "parentId": 2,
        "propertyId": 5,
        "selected": false,
        "value": "International schools",
      },
      Object {
        "group": "School system",
        "iconComponent": "icon-education",
        "id": 8,
        "name": "Dutch Schools",
        "parentId": 2,
        "propertyId": 6,
        "selected": false,
        "value": "Dutch schools",
      },
    ],
    "iconComponent": "icon-education",
    "isGroup": true,
    "name": "School system",
  },
]
`;

exports[`ComponentFilters should create 1`] = `
Array [
  Object {
    "iconComponent": "icon-education",
    "id": 2,
    "isFilter": true,
    "name": "Schools",
    "parentId": 0,
    "propertyId": null,
    "selected": false,
    "value": "School",
  },
  Object {
    "iconComponent": "icon-bus",
    "id": 1,
    "isFilter": true,
    "name": "Stations",
    "parentId": 0,
    "propertyId": null,
    "selected": false,
    "value": "Station",
  },
]
`;
